katz/src/main/kotlin/katz/typeclasses/Foldable.kt

```diff
@@ -0,0 +1,205 @@
+
+/**
+ * Data structures that can be folded to a summary value.
+ *
+ * Foldable<F> is implemented in terms of two basic methods:
+ *
+ *  - `foldLeft(fa, b)(f)` eagerly folds `fa` from left-to-right.
+ *  - `foldRight(fa, b)(f)` lazily folds `fa` from right-to-left.
+ *
+ * Beyond these it provides many other useful methods related to folding over F<A> values.
+ */
+interface Foldable<F> {
+
+    /**
+     * Left associative fold on F using the provided function.
+     */
+    fun <A, B> foldL(fa: HK<F, A>, b: B): ((B, A) -> B) -> B
+
+    /**
+     * Right associative lazy fold on F using the provided function.
+     *
+     * This method evaluates lb lazily (in some cases it will not be needed), and returns a lazy value. We are using
+     * (A, Eval<B>) => Eval<B> to support laziness in a stack-safe way. Chained computation should be performed via
+     * .map and .flatMap.
+     *
+     * For more detailed information about how this method works see the documentation for Eval<A>.
+     */
+    fun <A, B> foldR(fa: HK<F, A>, lb: Eval<B>): ((A, Eval<B>) -> Eval<B>) -> Eval<B>
+
+    /**
+     * The size of this Foldable.
+     *
+     * This can be overriden in structures that have more efficient size implementations.
+     *
+     * Note: will not terminate for infinite-sized collections.
+     */
+    fun <A> size(ml: Monoid<Long>, fa: HK<F, A>): Long = foldMap(ml, fa)({ _ -> 1L })
+
+    /**
+     * Fold implemented using the given Monoid<A> instance.
+     */
+    fun <A> fold(ma: Monoid<A>, fa: HK<F, A>): A =
+            foldL(fa, ma.empty())({ acc, a -> ma.combine(acc, a) })
+
+    /**
+     * Alias for fold.
+     */
+    fun <A : Monoid<*>> combineAll(m: Monoid<A>, fa: HK<F, A>): A = fold(m, fa)
+
+    /**
+     * Fold implemented by mapping A values into B and then combining them using the given Monoid<B> instance.
+     */
+    fun <A, B> foldMap(mb: Monoid<B>, fa: HK<F, A>): (f: (A) -> B) -> B =
+            { f: (A) -> B -> foldL(fa, mb.empty())({ b, a -> mb.combine(b, f(a)) }) }
+
+    /**
+     * Left associative monadic folding on F.
+     *
+     * The default implementation of this is based on foldL, and thus will always fold across the entire structure.
+     * Certain structures are able to implement this in such a way that folds can be short-circuited (not traverse the
+     * entirety of the structure), depending on the G result produced at a given step.
+     */
+    fun <G, A, B> foldM(g: Monad<G>, fa: HK<F, A>, z: B): ((B, A) -> HK<G, B>) -> HK<G, B> {
+        val foldL = foldL(fa, g.pure(z))
+        return { f: (B, A) -> HK<G, B> ->
+            foldL { gb, a -> g.flatMap(gb) { f(it, a) } }
+        }
+    }
+
+    /**
+     * Monadic folding on F by mapping A values to G<B>, combining the B values using the given Monoid<B> instance.
+     *
+     * Similar to foldM, but using a Monoid<B>.
+     */
+    fun <G, A, B> foldMapM(g: Monad<G>, bb: Monoid<B>, fa: HK<F, A>): ((A) -> HK<G, B>) -> HK<G, B> {
```

Comment:
If I remember correctly the current nomenclature we were using required for Monad instances to be capitalized. I may need to pop a link in here from another PR to support this statement :D.